src/main/java/com/squareup/javapoet/NameAllocator.java

```diff
@@ -0,0 +1,131 @@
+ *
+ * <p>One we've allocated names we can use them when generating code: <pre>   {@code
+ *
+ *   MethodSpec.Builder builder = MethodSpec.methodBuilder("toString")
+ *       .addAnnotation(Override.class)
+ *       .addModifiers(Modifier.PUBLIC)
+ *       .returns(String.class);
+ *
+ *   builder.addStatement("$1T $2N = new $1T()",
+ *   StringBuilder.class, nameAllocator.get("string builder"));
+ *   for (MyProperty property : properties) {
+ *     builder.addStatement("$N.append($N)",
+ *         nameAllocator.get("string builder"), nameAllocator.get(property));
+ *   }
+ *   builder.addStatement("return $N", nameAllocator.get("string builder"));
+ *   return builder.build();
+ * }</pre>
+ *
+ * The above code generates unique names if presented with conflicts. Given user-supplied properties
+ * with names {@code ab} and {@code sb} this generates the following:  <pre>   {@code
+ *
+ *   @Override
+ *   public String toString() {
+ *     StringBuilder sb_ = new StringBuilder();
+ *     sb_.append(ab);
+ *     sb_.append(sb);
+ *     return sb_.toString();
+ *   }
+ * }</pre>
+ *
+ * The underscore is appended to {@code sb} to avoid conflicting with the user-supplied {@code sb}
+ * property. Underscores are also prefixed for names that start with a digit, and used to replace
+ * name-unsafe characters like space or dash.
+ */
+public final class NameAllocator {
+  public static final Set<String> JAVA_KEYWORDS = Util.immutableSet(Arrays.asList(
+      "abstract", "assert", "boolean", "break", "byte", "case", "catch", "char",
+      "class", "const", "continue", "default", "do", "double", "else", "enum", "extends",
+      "final", "finally", "float", "for", "goto", "if", "implements", "import", "instanceof",
+      "int", "interface", "long", "native", "new", "package", "private", "protected", "public",
+      "return", "short", "static", "strictfp", "super", "switch", "synchronized", "this",
+      "throw", "throws", "transient", "try", "void", "volatile", "while"));
+
+  private final Set<String> allocatedNames = new LinkedHashSet<>(JAVA_KEYWORDS);
+  private final Map<Object, String> tagToName = new LinkedHashMap<>();
+
+  public String newName(String suggestion, Object tag) {
+    checkNotNull(suggestion, "suggestion");
+    checkNotNull(tag, "tag");
+
+    suggestion = toJavaIdentifier(suggestion);
+
+    while (!allocatedNames.add(suggestion)) {
+      suggestion = suggestion + "_";
+    }
+
+    String replaced = tagToName.put(tag, suggestion);
+    if (replaced != null) {
+      tagToName.put(tag, replaced); // Put things back as they were!
+      throw new IllegalArgumentException("tag " + tag + " cannot be used for both '" + replaced
+          + "' and '" + suggestion + "'");
+    }
+
+    return suggestion;
+  }
+
+  public static String toJavaIdentifier(String suggestion) {
+    StringBuilder result = new StringBuilder();
+    for (int i = 0, codePoint; i < suggestion.length(); i += Character.charCount(codePoint)) {
```

Comment:
I think this would be clearer as a `while` loop, or a `for` with an empty third part. Certainly the `codePoint` declaration without an initializer brought me up short.